### DISCUSSION TITLE
Failing to simulate a coaxial monopol

### DISCUSSION THREAD

**Discussion by gibbi** [thliebig/openEMS-Project#149] [General]
I am completely new to EMS and want to use it for Sleeve Dipole Simulations. I Modified the Patch antenna example to just an coax with a stripped end. I seem to be unable to excite the source as there seems to be no energy in the simulation. Can someone point me in the direction what i am doing wrong? 

Thank you so much

[ems_test.txt](https://github.com/thliebig/openEMS-Project/files/13401704/ems_test.txt)

**Comment by LubomirJagos** [comment]
maybe, but without source code it's like guessing from crystal ball :D what about to put here your code? just geuss, you done meshing wrong or ... put here your code, it's fastes solution

**Comment by gibbi** [comment]
The code is in the attached Textfile, I'd did not want to spam a big code comment :)

**Reply by LubomirJagos** [reply]
you were missing gridline in Z coord, also I put simulation output into current folder into ./simulation_output
the result S11 is -0.3 what is not good, so you can play with it, now this simulation is running

```python
import os, tempfile
from pylab import *

from CSXCAD  import ContinuousStructure
from CSXCAD import AppCSXCAD_BIN
from openEMS import openEMS
from openEMS.physical_constants import *

### Create folder if not exists in current folder
current_directory = os.getcwd()
Sim_Path = os.path.join(current_directory, r'simulation_output')
if not os.path.exists(Sim_Path):
   os.makedirs(Sim_Path)

post_proc_only = False		#THIS CAN BLOCK openEMS calculate simulation

feed_R = 50     #feed resistance

# size of the simulation box
SimBox = np.array([100, 100, 300])

# setup FDTD parameter & excitation function
f0 = 2e9 # center frequency
fc = 1e9 # 20 dB corner frequency

### FDTD setup
## * Limit the simulation to 30k timesteps
## * Define a reduced end criteria of -40dB
FDTD = openEMS(NrTS=30000, EndCriteria=1e-4)
FDTD.SetGaussExcite( f0, fc )
FDTD.SetBoundaryCond( ['PML_8', 'PML_8', 'PML_8', 'PML_8', 'PML_8', 'PML_8'] )


CSX = ContinuousStructure()
FDTD.SetCSX(CSX)
mesh = CSX.GetGrid()
mesh.SetDeltaUnit(1e-3)
mesh_res = C0/(f0+fc)/1e-3/40

### Generate properties, primitives and mesh-grid
#initialize the mesh with the "air-box" dimensions
mesh.AddLine('x', [-SimBox[0]/2, SimBox[0]/2])
mesh.AddLine('y', [-SimBox[1]/2, SimBox[1]/2])
mesh.AddLine('z', [-SimBox[2]/2, SimBox[2]/2])

mesh.SmoothMeshLines('all', mesh_res, 1.4)

gnd = CSX.AddMetal( 'gnd' ) # create a perfect electric conductor (PEC)

centerConductorRadius = 0.94/2
ptfeRadius = 2.95/2
outerShieldRadius = 3.58/2

centerConductor = CSX.AddMetal( 'centerConductor' )
start = [0, 0, 0]
stop  = [0, 0, 100]
centerConductor.AddCylinder(priority=10, start=start, stop=stop, radius=centerConductorRadius)
FDTD.AddEdges2Grid(dirs='xy', properties=centerConductor, metal_edge_res=mesh_res/4)

start = [0, 0, 0]
stop  = [0, 0, 50]
width = (outerShieldRadius-ptfeRadius)
gnd.AddCylindricalShell(priority=10, start=start, stop=stop, radius=ptfeRadius+width/2, shell_width=width)
FDTD.AddEdges2Grid(dirs='xy', properties=gnd, metal_edge_res=mesh_res/4)

# create patch
ptfe = CSX.AddMaterial( 'PTFE', epsilon=2, kappa=0.124)
start = [0, 0, 0]
stop  = [0, 0, 50]
width = (ptfeRadius-centerConductorRadius)
ptfe.AddCylindricalShell(priority=10, start=start, stop=stop, radius=centerConductorRadius+width/2, shell_width=width)

# apply the excitation & resist as a current source
#
# make excitation thicker not just line to be sure there are some gridlines included
#
start = [0, 0.2, 0.2]
stop  = [outerShieldRadius, -0.2, -0.2]
port = FDTD.AddLumpedPort(1, feed_R, start, stop, 'x', 1.0, priority=5, edges2grid='xy')
mesh.AddLine('z', [0])       #add line in Z to go through excitation

Et = CSX.AddDump('Et', file_type=0)
start = [-50, -50, 0];
stop  = [50, 50, 150];
Et.AddBox(start, stop);

### Run the simulation
CSX_file = os.path.join(Sim_Path, 'monopol.xml')
if not os.path.exists(Sim_Path):
    os.mkdir(Sim_Path)
CSX.Write2XML(CSX_file)

os.system(AppCSXCAD_BIN + ' "{}"'.format(CSX_file))

if not post_proc_only:
    FDTD.Run(Sim_Path, verbose=3, cleanup=True)

# plot s11
### Post-processing and plotting
f = np.linspace(max(1e9,f0-fc),f0+fc,401)
port.CalcPort(Sim_Path, f)

print(port.uf_ref,port.uf_inc)
s11 = port.uf_ref/port.uf_inc
s11_dB = 20.0*np.log10(np.abs(s11))
figure()
plot(f/1e9, s11_dB, 'k-', linewidth=2, label='$S_{11}$')
grid()
legend()
ylabel('S-Parameter (dB)')
xlabel('Frequency (GHz)')
show()
```

**Reply by Sarajevo67** [reply]
Is it possible to have some kind of 'face' excitations for monopoles and dipoles? Rectangular in Cartesian but would be best in Cylindrical coordinate system?

**Reply by LubomirJagos** [reply]
@Sarajevo67 yes, they are there, this ones:
  - AddCircWaveGuidePort.m
  - AddRectWaveGuidePort.m
  - AddPlaneWaveExcite.m
  
  They are defined in matlab/octave interface, in core what they are doing for excitation you are able to define weighting function for X,Y,Z so these ones just already have built in calculations for mode TE10, TE11, ... 
  All of them are inside using AddExcitation function:

this is from AddCircWaveGuidePort.m
```matlab
func_Er = [ num2str(-1/kc_draw^2,15) '/rho*cos(' angle ')*j1('  num2str(kc_draw,15) '*rho)'];
func_Ea = [ num2str(1/kc_draw,15) '*sin(' angle ')*0.5*(j0('  num2str(kc_draw,15) '*rho)-jn(2,'  num2str(kc_draw,15) '*rho))'];

% magnetic field mode profile
func_Hr = [ num2str(-1/kc_draw,15) '*sin(' angle ')*0.5*(j0('  num2str(kc_draw,15) '*rho)-jn(2,'  num2str(kc_draw,15) '*rho))'];
func_Ha = [ num2str(-1/kc_draw^2,15) '/rho*cos(' angle ')*j1('  num2str(kc_draw,15) '*rho)'];

if (CSX.ATTRIBUTE.CoordSystem==1)
    func_E = {func_Er, func_Ea, 0};
    func_H = {func_Hr, func_Ha, 0};
else
    func_Ex = ['(' func_Er '*cos(a) - ' func_Ea '*sin(a) ) * (rho<' num2str(radius/unit) ')'];
    func_Ey = ['(' func_Er '*sin(a) + ' func_Ea '*cos(a) ) * (rho<' num2str(radius/unit) ')'];
    func_E = {func_Ex, func_Ey, 0};
    
    func_Hx = ['(' func_Hr '*cos(a) - ' func_Ha '*sin(a) ) * (rho<' num2str(radius/unit) ')'];
    func_Hy = ['(' func_Hr '*sin(a) + ' func_Ha '*cos(a) ) * (rho<' num2str(radius/unit) ')'];
    func_H = {func_Hx, func_Hy, 0};
end

[CSX,port] = AddWaveGuidePort( CSX, prio, portnr, start, stop, 2, func_E, func_H, kc, exc_amp, varargin{:} );
````

this is from AddWaveGuidePort.m
```matlab
    e_start = start;
    e_stop = stop;
    e_stop(dir) = e_start(dir);
    port.excite = 1;
    port.excitepos = e_start(dir);
    e_vec = [1 1 1]*exc_amp;
    e_vec(dir) = 0;
    exc_name = [PortNamePrefix 'port_excite_' num2str(portnr)];
    CSX = AddExcitation( CSX, exc_name, 0, e_vec, varargin{:});
    CSX = SetExcitationWeight(CSX, exc_name, E_WG_func );
    CSX = AddBox( CSX, exc_name, prio, e_start, e_stop); 
```
What I was briefly looking they are not part of python interface so it's needed to rewrite them into python functions.

**Reply by LubomirJagos** [reply]
that matlab/octave interface is more mature and there is more stuff done which makes simulations more easy, I think it was first developed and there was most work done and then Thorsten create python interface and interface directly C++ function, therefore this high level stuff which is pure matlab/octave stays not available and need to be rewritten into python
but what is great is that since there is that matlab interface you have reference to validate it so brute force CTRL+C, CTRL+V and validate results agains matlab/octave interace is way how to do it and it will be right.

Also in AddCircWaveGuidePort.m there is error as now it's just generate circ port when propagation direction is in Z axis, I have to modified for cartesian to be able have circ waveguide also in X or Y direction, it's just swapping equation for X,Y,Z.

**Reply by Sarajevo67** [reply]
> it's just generate circ port when propagation direction is in Z axis,

It is not problem, because monopoles are usually modeled in z direction.

So these waveguide excitation ports, specifically AddCircWaveGuidePort, cold be used to 'face' excite cylinder monopol?

I guess that start, stop and radius are aligned to cylinder face, without oversizing?

**Reply by LubomirJagos** [reply]
In python there is just wave guide port function, there is no circular port, therefore you have to write E mode function yourself
https://docs.openems.de/python/openEMS/ports.html

Myself I was not trying this, I need to look on equations for coaxial cable, but just shortly it's TEM mode what is E and H field mode so there must be both weighting functions for some plane perpendicular to propagation direction defined.

Anyway if somebody done this in python it can be wrapped into high level python interace function which is now missing.

**Reply by LubomirJagos** [reply]
this comment here is just to specified what we are talking about, there are "modes" which propagates through waveguide, they are solutions of Maxwell equation for specific gemoetric structure, each of them need some breaking limit energy, they can be ordered from lowest to highest energy what is the braking point when input excitation has equal or higher energy at input frequency
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/fd9b5bec-95a0-416b-951e-36070dedc87a)
src: https://en.wikipedia.org/wiki/Alternating_current

the field will form this shape inside structure and propagates through input to output of structure, information is then modulated on this field like fluctuation of it's amplitude
these modes are well known for cylindrical or rectangular shapes (waveguides), here is picture of them
![800px-Selected_modes svg](https://github.com/thliebig/openEMS-Project/assets/18229825/7cf8552b-a6db-49eb-89ce-e21b6b34b52c)
src: https://en.wikipedia.org/wiki/Transverse_mode

**Reply by LubomirJagos** [reply]
if you are wondering "yes" this is the same stuff what electrons are doing inside atom cores, since they are electric charged they formed modes which are better known as orbitals, these orbitals have to also fulfill Maxwell equations, I'm not expert but this is just for information if somebody was wandering that why "modes" and electron orbitals seems similar, it's due they are solution of the same equations
solution means it's some function which has space distribution of EM field which is well defined on borders (0 or maximum) in case of waveguides there is 0V on edge since it's ground and maximum for magnetic field
it's same theory as musical instruments, since every guitar, flute or whatever where air resonates at some frequency creates modes which are specified by instrument shape, it's possible field distribution which is able to exists based on mathematic boundary condition and sturcutrue shape inside
so yes all EM theory, music, water pipes theory are same since they are working with math for fields and all fields have the same mathematic equations

**Reply by LubomirJagos** [reply]
using lumped port for this simulation is right by me since it's real case, it's same as put directly output from some pin of processor to coax cable
the case when putting waveguide mode is the same as to put balun to coax cable, if I understand it correctly balun is the device which convert field from PCB to some structure it's not just impedance transformer but also field transformer, so it's transition between one field shape to another field shape, on PCB there is assymetric field (from active trace to gnd) and coax cable or waveguide are symmetric ones, therefore using waveguide port is right in sense there is balun, but using lumped port is same case as directly wiring pcb trace to coax
both of these cases are right simulations but results are different :)

I'm not saying that it's right to put PCB output directly to caox, it's just possible and it will somehow run :D and what is showing openEMS it's not preferable way to do things :D

**Reply by Sarajevo67** [reply]
According to: **A Practical Guide to 3-D Simulation by Thomas Weiland, Martin Timm, and Irina Munteanu,** 
size of waveguide port for coaxial excitation is identical to the inner cylindrical shell diameter. If analogy with MSL applies (Figure 10.), where width of waveguide port is 6-10 MSL widths and height is 5 substrate heights, monopole/dipole excitation WG port would be 5-10 monopole/dipole cylinder diameters?

**EDIT: I perfectly understand that this is overkill for everyday simulation.**

**Reply by LubomirJagos** [reply]
aaaaah I remember from university when we were modelling few years ago patch antenna that professor mentioned this rule and we use for input microwave port in HFSS port which was 10x more width then input microstrip and 5x times more in height as subtrate, I think this is case when you put there waveguide which models planar wave, but I'm not really sure if this is case of these waveguides, since what I imagine is it's like direct pipes fittings and you defines equations for that port which has EXACT geometry, so I my guess is to give that port same diemnsions as that input port has and define E weighting equation (I was getting error that openEMS doesn't support H equation for weights) hmmmm now writing don't know really how, I will rewrite that octave one and create example for cylindrical and rectangular horn and make like 10 simulations and looks for results

**Reply by LubomirJagos** [reply]
problem with HFSS is that HFSS is not FDTD, it's FEM, so it performs multifrequency simulations in way that it's doing multirans, since it's FEM it's meshing for freqeuncy and doing calcualtion at exact frequency what is much much faster but it's not accurate for ultrabandwidth, FDTD is like ultrabandwidth what is slow but accurate, depends, and this rule hmmmmmm have to think and experiment little, what you are saying is right but that's case for planar wave, it's the case as there is some port and you are lighting on it from output from some horn antenna, I imagine it always like the wave like our sun is ligthing on us, that we are too far and since then all light is coherent (no phase difeerence for points on edges) , but I'm not sure if this is that case

**Reply by LubomirJagos** [reply]
this is full AddCircWaveGuide.m
```matlab
function [CSX,port] = AddCircWaveGuidePort( CSX, prio, portnr, start, stop, radius, mode_name, pol_ang, exc_amp, varargin )
% function [CSX,port] = AddCircWaveGuidePort( CSX, prio, portnr, start, stop, radius, mode_name, pol_ang, exc_amp, varargin )
% 
% Create a circular waveguide port, including an optional excitation and probes
% 
% Note: - The excitation will be located at the start position in the given direction
%       - The voltage and current probes at the stop position in the given direction
%
% input:
%   CSX:        complete CSX structure (must contain a mesh)
%   prio:       priority of primitives
%   start:      start coordinates of waveguide port box
%   stop:       stop  coordinates of waveguide port box
%   radius:     circular waveguide radius (in meter)
%   mode_name:  mode name, e.g. 'TE11' or 'TM21'
%   pol_ang:    polarization angle (e.g. 0 = horizontal, pi/2 = vertical)
%   exc_amp:    excitation amplitude (set 0 to be passive)
%
% optional (key/values):
%   varargin:   optional additional excitations options, see also AddExcitation
%   'PortNamePrefix': a prefix to the port name
%
% output:
%   CSX:        modified CSX structure
%   port:       port structure to use with calcPort
%
% example:
%   % create a TE11 circular waveguide mode, using cylindircal coordinates
%   start=[mesh.r(1)   mesh.a(1)   0  ];
%   stop =[mesh.r(end) mesh.a(end) 100];
%   [CSX,port] = AddCircWaveGuidePort( CSX, 99, 1, start, stop, 320e-3, 'TE11', 0, 1);
%
% openEMS matlab interface
% -----------------------
% (c) 2013 Thorsten Liebig (thorsten.liebig@gmx.de)
%
% See also InitCSX, AddExcitation, calcWGPort, calcPort

if (~strcmpi(mode_name(1:2),'TE'))
    error 'currently only TE type modes are supported'
end

if (nargin<9)
    exc_amp = 0;
end
if (nargin<8)
    pol_ang = 0;
end

pnm = 0;
n = str2double(mode_name(3));
m = str2double(mode_name(4));

% values by David M. Pozar, Microwave Engineering, third edition
if ((n==0) && (m==1))
    pnm = 3.832;
elseif ((n==1) && (m==1))
    pnm = 1.841;
elseif ((n==2) && (m==1))
    pnm = 3.054;
elseif ((n==0) && (m==2))
    pnm = 7.016;
elseif ((n==1) && (m==2))
    pnm = 5.331;
elseif ((n==2) && (m==2))
    pnm = 6.706;
elseif ((n==0) && (m==3))
    pnm = 10.174;
elseif ((n==1) && (m==3))
    pnm = 8.536;
elseif ((n==2) && (m==3))
    pnm = 9.970;
else
    error 'invalid TE_nm mode'
end

if ~isfield(CSX,'RectilinearGrid')
    error 'mesh needs to be defined! Use DefineRectGrid() first!';
end

unit = CSX.RectilinearGrid.ATTRIBUTE.DeltaUnit;
kc = pnm/radius;
kc_draw = kc*unit;

angle = ['a-' num2str(pol_ang)];
% functions by David M. Pozar, Microwave Engineering, third edition
% electric field mode profile
func_Er = [ num2str(-1/kc_draw^2,15) '/rho*cos(' angle ')*j1('  num2str(kc_draw,15) '*rho)'];
func_Ea = [ num2str(1/kc_draw,15) '*sin(' angle ')*0.5*(j0('  num2str(kc_draw,15) '*rho)-jn(2,'  num2str(kc_draw,15) '*rho))'];

% magnetic field mode profile
func_Hr = [ num2str(-1/kc_draw,15) '*sin(' angle ')*0.5*(j0('  num2str(kc_draw,15) '*rho)-jn(2,'  num2str(kc_draw,15) '*rho))'];
func_Ha = [ num2str(-1/kc_draw^2,15) '/rho*cos(' angle ')*j1('  num2str(kc_draw,15) '*rho)'];

if (CSX.ATTRIBUTE.CoordSystem==1)
    func_E = {func_Er, func_Ea, 0};
    func_H = {func_Hr, func_Ha, 0};
else
    func_Ex = ['(' func_Er '*cos(a) - ' func_Ea '*sin(a) ) * (rho<' num2str(radius/unit) ')'];
    func_Ey = ['(' func_Er '*sin(a) + ' func_Ea '*cos(a) ) * (rho<' num2str(radius/unit) ')'];
    func_E = {func_Ex, func_Ey, 0};
    
    func_Hx = ['(' func_Hr '*cos(a) - ' func_Ha '*sin(a) ) * (rho<' num2str(radius/unit) ')'];
    func_Hy = ['(' func_Hr '*sin(a) + ' func_Ha '*cos(a) ) * (rho<' num2str(radius/unit) ')'];
    func_H = {func_Hx, func_Hy, 0};
end

[CSX,port] = AddWaveGuidePort( CSX, prio, portnr, start, stop, 2, func_E, func_H, kc, exc_amp, varargin{:} );
```

just briefly this is most importnat piece of code:
```matalb
unit = CSX.RectilinearGrid.ATTRIBUTE.DeltaUnit;
kc = pnm/radius;
kc_draw = kc*unit;
```

by my opinion what it's doing it's scaling wave equations to port dimensions, SO NO NO MAKING PORT BIGGER, FIT IT TO YOUR INPUT PORT and wave weights are scaled to your input geometry

**Reply by LubomirJagos** [reply]
David M. Pozar, Microwave Engineering is like microwave Bible, there is also that guy from Greece Orfanidis for antennas and signal processing :D

**Comment by LubomirJagos** [comment]
ok, there was lot's of talking, you need answer, so here it is:
    - symmetric excitation done for coax using 2 lumped port (I took this from how matlab/octave generate coax excitation using 2 lumped port excitation to excite coax symmetric)
    - added ground plane for your monopole
    
s11
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/f863165e-f4db-4a2e-87c8-beffaaf66398)

symmetric excitation using 2 lumped ports
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/55c336f5-cc00-40fb-a076-bb94fa61ddff)

whole geometry
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/be7d5fd3-ba65-46a0-a9ec-0fc0cee6f2a8)

simulation script
```matlab
import os, tempfile
from pylab import *

from CSXCAD  import ContinuousStructure
from CSXCAD import AppCSXCAD_BIN
from openEMS import openEMS
from openEMS.physical_constants import *

### Create folder if not exists in current folder
current_directory = os.getcwd()
Sim_Path = os.path.join(current_directory, r'simulation_output')
if not os.path.exists(Sim_Path):
   os.makedirs(Sim_Path)

post_proc_only = False		#THIS CAN BLOCK openEMS calculate simulation

feed_R = 50     #feed resistance

# size of the simulation box
SimBox = np.array([100, 100, 300])

# setup FDTD parameter & excitation function
f0 = 4e9 # center frequency
fc = 3e9 # 20 dB corner frequency

### FDTD setup
## * Limit the simulation to 30k timesteps
## * Define a reduced end criteria of -40dB
FDTD = openEMS(NrTS=30000, EndCriteria=1e-4)
FDTD.SetGaussExcite( f0, fc )
FDTD.SetBoundaryCond( ['PML_8', 'PML_8', 'PML_8', 'PML_8', 'PML_8', 'PML_8'] )


CSX = ContinuousStructure()
FDTD.SetCSX(CSX)
mesh = CSX.GetGrid()
mesh.SetDeltaUnit(1e-3)
mesh_res = C0/(f0+fc)/1e-3/40

### Generate properties, primitives and mesh-grid
#initialize the mesh with the "air-box" dimensions
mesh.AddLine('x', [-SimBox[0]/2, SimBox[0]/2])
mesh.AddLine('y', [-SimBox[1]/2, SimBox[1]/2])
mesh.AddLine('z', [-SimBox[2]/2, SimBox[2]/2])

mesh.SmoothMeshLines('all', mesh_res, 1.4)

gnd = CSX.AddMetal( 'gnd' ) # create a perfect electric conductor (PEC)

centerConductorRadius = 0.94/2
ptfeRadius = 2.95/2
outerShieldRadius = 3.58/2

centerConductor = CSX.AddMetal( 'centerConductor' )
start = [0, 0, 0]
stop  = [0, 0, 100]
centerConductor.AddCylinder(priority=10, start=start, stop=stop, radius=centerConductorRadius)
FDTD.AddEdges2Grid(dirs='xy', properties=centerConductor, metal_edge_res=mesh_res/4)

start = [0, 0, 0]
stop  = [0, 0, 50]
width = (outerShieldRadius-ptfeRadius)
gnd.AddCylindricalShell(priority=10, start=start, stop=stop, radius=ptfeRadius+width/2, shell_width=width)
FDTD.AddEdges2Grid(dirs='xy', properties=gnd, metal_edge_res=mesh_res/4)

#create gnd
start = [0, 0, 50]
stop = [0, 0, 51]
gnd.AddCylindricalShell(priority=20, start=start, stop=stop, radius=20, shell_width=37)

# create patch
ptfe = CSX.AddMaterial( 'PTFE', epsilon=2, kappa=0.124)
start = [0, 0, 0]
stop  = [0, 0, 50]
width = (ptfeRadius-centerConductorRadius)
ptfe.AddCylindricalShell(priority=10, start=start, stop=stop, radius=centerConductorRadius+width/2, shell_width=width)

# apply the excitation & resist as a current source
#
# make excitation thicker not just line to be sure there are some gridlines included
#
start = [0, 0.2, 0.2]
stop  = [outerShieldRadius, -0.2, -0.2]
port = FDTD.AddLumpedPort(1, feed_R, start, stop, 'x', 1.0, priority=5, edges2grid='xy')

start = [0, 0.2, 0.2]
stop  = [-outerShieldRadius, -0.2, -0.2]
port = FDTD.AddLumpedPort(2, feed_R, start, stop, 'x', 1.0, priority=5, edges2grid='xy')

mesh.AddLine('z', [0])       #add line in Z to go through excitation

Et = CSX.AddDump('Et', file_type=0)
start = [-50, -50, 0];
stop  = [50, 50, 150];
Et.AddBox(start, stop);

### Run the simulation
CSX_file = os.path.join(Sim_Path, 'monopol.xml')
if not os.path.exists(Sim_Path):
    os.mkdir(Sim_Path)
CSX.Write2XML(CSX_file)

os.system(AppCSXCAD_BIN + ' "{}"'.format(CSX_file))

if not post_proc_only:
    FDTD.Run(Sim_Path, verbose=3, cleanup=True)

# plot s11
### Post-processing and plotting
f = np.linspace(max(1e9,f0-fc),f0+fc,401)
port.CalcPort(Sim_Path, f)

print(port.uf_ref,port.uf_inc)
s11 = port.uf_ref/port.uf_inc
s11_dB = 20.0*np.log10(np.abs(s11))
figure()
plot(f/1e9, s11_dB, 'k-', linewidth=2, label='$S_{11}$')
grid()
legend()
ylabel('S-Parameter (dB)')
xlabel('Frequency (GHz)')
show()
```

**Comment by LubomirJagos** [comment]
this is simulation without ground plane
    - no gnd plane
    - symmetric excitation

s11
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/c55857d9-cef8-488a-bf13-ef5e6cb66f7b)

gemoetry
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/90fb89eb-0125-49f6-a9e5-6d713d91d3ba)

code
```matlab
import os, tempfile
from pylab import *

from CSXCAD  import ContinuousStructure
from CSXCAD import AppCSXCAD_BIN
from openEMS import openEMS
from openEMS.physical_constants import *

### Create folder if not exists in current folder
current_directory = os.getcwd()
Sim_Path = os.path.join(current_directory, r'simulation_output')
if not os.path.exists(Sim_Path):
   os.makedirs(Sim_Path)

post_proc_only = False		#THIS CAN BLOCK openEMS calculate simulation

feed_R = 50     #feed resistance

# size of the simulation box
SimBox = np.array([100, 100, 300])

# setup FDTD parameter & excitation function
f0 = 3e9 # center frequency
fc = 2.5e9 # 20 dB corner frequency

### FDTD setup
## * Limit the simulation to 30k timesteps
## * Define a reduced end criteria of -40dB
FDTD = openEMS(NrTS=30000, EndCriteria=1e-4)
FDTD.SetGaussExcite( f0, fc )
FDTD.SetBoundaryCond( ['PML_8', 'PML_8', 'PML_8', 'PML_8', 'PML_8', 'PML_8'] )


CSX = ContinuousStructure()
FDTD.SetCSX(CSX)
mesh = CSX.GetGrid()
mesh.SetDeltaUnit(1e-3)
mesh_res = C0/(f0+fc)/1e-3/40

### Generate properties, primitives and mesh-grid
#initialize the mesh with the "air-box" dimensions
mesh.AddLine('x', [-SimBox[0]/2, SimBox[0]/2])
mesh.AddLine('y', [-SimBox[1]/2, SimBox[1]/2])
mesh.AddLine('z', [-SimBox[2]/2, SimBox[2]/2])

mesh.SmoothMeshLines('all', mesh_res, 1.4)

gnd = CSX.AddMetal( 'gnd' ) # create a perfect electric conductor (PEC)

centerConductorRadius = 0.94/2
ptfeRadius = 2.95/2
outerShieldRadius = 3.58/2

centerConductor = CSX.AddMetal( 'centerConductor' )
start = [0, 0, 0]
stop  = [0, 0, 100]
centerConductor.AddCylinder(priority=10, start=start, stop=stop, radius=centerConductorRadius)
FDTD.AddEdges2Grid(dirs='xy', properties=centerConductor, metal_edge_res=mesh_res/4)

start = [0, 0, 0]
stop  = [0, 0, 50]
width = (outerShieldRadius-ptfeRadius)
gnd.AddCylindricalShell(priority=10, start=start, stop=stop, radius=ptfeRadius+width/2, shell_width=width)
FDTD.AddEdges2Grid(dirs='xy', properties=gnd, metal_edge_res=mesh_res/4)

#create gnd
#start = [0, 0, 50]
#stop = [0, 0, 51]
#gnd.AddCylindricalShell(priority=20, start=start, stop=stop, radius=20, shell_width=37)

# create patch
ptfe = CSX.AddMaterial( 'PTFE', epsilon=2, kappa=0.124)
start = [0, 0, 0]
stop  = [0, 0, 50]
width = (ptfeRadius-centerConductorRadius)
ptfe.AddCylindricalShell(priority=10, start=start, stop=stop, radius=centerConductorRadius+width/2, shell_width=width)

# apply the excitation & resist as a current source
#
# make excitation thicker not just line to be sure there are some gridlines included
#
start = [0, 0.2, 0.2]
stop  = [outerShieldRadius, -0.2, -0.2]
port = FDTD.AddLumpedPort(1, feed_R, start, stop, 'x', 1.0, priority=5, edges2grid='xy')

start = [0, 0.2, 0.2]
stop  = [-outerShieldRadius, -0.2, -0.2]
port = FDTD.AddLumpedPort(2, feed_R, start, stop, 'x', 1.0, priority=5, edges2grid='xy')

mesh.AddLine('z', [0])       #add line in Z to go through excitation

Et = CSX.AddDump('Et', file_type=0)
start = [-50, -50, 0];
stop  = [50, 50, 150];
Et.AddBox(start, stop);

### Run the simulation
CSX_file = os.path.join(Sim_Path, 'monopol.xml')
if not os.path.exists(Sim_Path):
    os.mkdir(Sim_Path)
CSX.Write2XML(CSX_file)

os.system(AppCSXCAD_BIN + ' "{}"'.format(CSX_file))

if not post_proc_only:
    FDTD.Run(Sim_Path, verbose=3, cleanup=True)

# plot s11
### Post-processing and plotting
f = np.linspace(max(1e9,f0-fc),f0+fc,401)
port.CalcPort(Sim_Path, f)

print(port.uf_ref,port.uf_inc)
s11 = port.uf_ref/port.uf_inc
s11_dB = 20.0*np.log10(np.abs(s11))
figure()
plot(f/1e9, s11_dB, 'k-', linewidth=2, label='$S_{11}$')
grid()
legend()
ylabel('S-Parameter (dB)')
xlabel('Frequency (GHz)')
show()
```

**Comment by gibbi** [comment]
first of all, WOW. Now that is one comprehensive response. Thanks you so much for going into so much detail - i can't wait to try out more. Now i can focus on the fun part of replicating our antenna.
